src/redux-middleware/remoteSync.ts

```diff
@@ -0,0 +1,121 @@
+ */
+const remoteSyncMiddleware: ThunkMiddleware<State> = ({ getState, dispatch }) => {
+
+  let thoughtListeners: Index<Ref> = {}
+  let lastExpanded: Index<Context> = {} // eslint-disable-line fp/no-let
+  let thoughts: (Lexeme | undefined)[] = []// eslint-disable-line fp/no-let
+
+  return next => async action => {
+
+    next(action)
+    if (action.type !== 'existingThoughtChange') return
+
+    const state = getState()
+
+    if (!state.authenticated || !state?.user?.uid) return
+
+    const expandedContexts = expandThoughts(state, state.cursor, {
+      returnContexts: true,
+    })
+
+    const visibleContexts = getVisibleContexts(state, expandedContexts)
+    const visibleContextIds = Object.keys(visibleContexts)
+
+    if (Object.keys(state.expanded).length > 0
+    && equalArrays(Object.keys(expandedContexts), Object.keys(lastExpanded))) return
+
+    lastExpanded = expandedContexts
+
+    const listenersToAdd = _.difference(visibleContextIds, Object.keys(thoughtListeners))
+    const listenersToRemove = _.difference(Object.keys(thoughtListeners), visibleContextIds)
+
+    const refsToAdd: Index<Ref> = listenersToAdd.reduce((acc, id) => {
+      const url = `users/${state.user!.uid}/contextIndex/${id}`
+      return { ...acc, [id]: window.firebase?.database().ref(url) }
+    }, {} as Index<Ref>)
+
+    const refsToRemove = listenersToRemove.reduce((acc, id) => {
+      return { ...acc, [id]: thoughtListeners[id] }
+    }, {} as Index<Ref>)
+    const provider = getFirebaseProvider(getState(), dispatch)
+
+    Object.keys(refsToRemove).forEach(id => {
+      refsToRemove[id]?.off('value')
+    })
+
+    Object.keys(refsToAdd).forEach(id => {
+
+      refsToAdd[id]?.on('value', _.debounce(async snapshot => {
+
+        const state = getState()
+
+        const remoteState: Parent = snapshot.val()
+
+        if (!remoteState || refsToRemove[id]) return
+
+        const contexts: Context[] = remoteState.children.map(child => [...remoteState.context, child.value])
+        const thoughtIds = contexts.map(context => hashThought(head(context)))
+
+        const thoughtPromises = thoughtIds.map(id => provider.getThoughtById(id))
+
+        thoughts = await Promise.all(thoughtPromises)
+        const thoughtindex: Index<Lexeme> = thoughts.reduce((acc, thought, i) => {
+          return { ...acc, ...thought ? { [thoughtIds[i]]: { ...thought, id: thoughtIds[i] } } : false }
+        }, {})
+
+        if (Object.keys(thoughtindex).length !== thoughtIds.length) {
+          // Couldn't fetch all thoughts
+          return
+        }
+
+        dispatch(reconcile({ thoughtsResults: [state.thoughts, {
+          thoughtIndex: thoughtindex,
+          contextIndex: { [id]: remoteState },
+          thoughtCache: [],
+          contextCache: []
+        }] }))
+
+      }, 500))
+
+    })
+
+    thoughtListeners = _.omit({ ...thoughtListeners, ...refsToAdd }, listenersToRemove)
```

Comment:
I'm on the fence about whether to allow this mutation. It's something we're really trying to avoid on this project.